File: src/shared/faults/delay.ts
const DEFAULT_DELAY_MS = Number(process.env.DEFAULT_DELAY_MS ?? 0);

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Simulates processing latency. Pass an explicit ms to override
 * per-call (e.g. from fault_config); otherwise falls back to
 * DEFAULT_DELAY_MS so latency is controllable without code changes.
 */
export async function simulateLatency(ms?: number): Promise<void> {
  const delay = ms ?? DEFAULT_DELAY_MS;
  if (delay > 0) {
    await sleep(delay);
  }
}

File: internal/shared/faults/delay.go
// Package faults provides small helpers for simulating processing latency.
package faults

import (
	"context"
	"time"
)

// defaultDelay is the fallback latency (DEFAULT_DELAY_MS) applied when a
// call to SimulateLatency doesn't pass an explicit override.
var defaultDelay time.Duration

// SetDefaultDelay configures the fallback latency used by SimulateLatency
// when no explicit delay is given. Called once at startup from config.
func SetDefaultDelay(d time.Duration) {
	defaultDelay = d
}

// SimulateLatency sleeps for d if positive, otherwise falls back to the
// configured default delay. Respects context cancellation.
func SimulateLatency(ctx context.Context, d *time.Duration) {
	delay := defaultDelay
	if d != nil {
		delay = *d
	}
	if delay <= 0 {
		return
	}
	select {
	case <-time.After(delay):
	case <-ctx.Done():
	}
}
